Allow sending interrupts and newline, and allow focusing the REPL

thezeroalpha opened the pull request:
I added a few modifications for my convenience; if you think these might be useful for all users, feel free to integrate this PR.

My changes:

* allow sending newline to the REPL: mapping `<Plug>ReplSendNewline`, default mapping `gz<CR>`
* allow sending ctrl-c (sigint) to the REPL: mapping `<Plug>ReplSendCtrlC`, default mapping `gzc`
* allow sending ctrl-d (eof) to the REPL: mapping `<Plug>ReplSendCtrlD`, default mapping `gzq`
* allow focusing the window containing the REPL from anywhere, opens the REPL in a new split window if it's not currently in a window: mapping `<Plug>ReplFocus`, default mapping `gz<TAB>`

The 'default' mappings are just what I use, I'm not sure if they're the best defaults for all users.

The reason I did not use the existing `zepl#send()` function is because 1) the function does some extra text processing, and 2) for some reason the special characters get interpreted and it doesn't work. If you have any ideas for how to make this work, please let me know.

Also, I am not very experienced with vimscript yet, so there are probbly better/more efficient ways of writing what I did.

thezeroalpha committed:
Allow sending interrupts, newline, and focusing

Default bindings: Focus the REPL with gz<tab>, send control-c with gzc,
send control-d with gzq, send newline with gz<cr>.

thezeroalpha committed:
Re-added two things I didn't mean to remove

axvr commented on autoload/zepl.vim:
```diff
@@ -100,6 +100,46 @@ function! zepl#send(text) abort
+  else
+    exe 'sbuffer '.s:repl_bufnr
+  endif
+endfunction
```
It looks like this is doing the same thing as `zepl#jump`, but I do prefer the name you chose (`zepl#focus`)

thezeroalpha replied:
Yes, it's similar. One of the reasons I made a separate function was because `zepl#jump` takes parameters, and I wasn't sure what to pass in to just focus the window -- now I see that `call zepl#jump('', 0)` would do the job.

However, there is one difference -- this function always focuses the active REPL, even if it's in a different tab (which I prefer), whereas `zepl#jump` opens the REPL in a new split.

axvr replied:
I've changed all of the public functions so now they have optional parameters where possible; which now means that `call zepl#jump()` with no parameters will work.

`sbuffer` (and `zepl#jump`) can actually focus buffers in other tabs, but you have to modify `:help 'switchbuf'`.

```vim
set switchbuf+=usetab
call zepl#jump()
```

thezeroalpha replied:
Awesome, thanks. Works just as expected. Since there is only one REPL at a time, I added `usetab` to the `set switchbuf` line [in autoload/zepl.vim](https://github.com/thezeroalpha/zepl.vim/blob/75b60eb8d5b315656c4267773028e2927f67b0bc/autoload/zepl.vim#L72), because it makes more sense to me to switch to the existing REPL (even if it's in a separate tab), as opposed to opening it in a split.

axvr replied:
I have actually considered making that the default option before, however if for whatever reason a user doesn't want that behaviour, they would be unable to turn it off as the option is always enabled as part of zepl.vim.

I definitely need to update the documentation to make this clear.

thezeroalpha replied:
Another thing you could potentially do is let the user set a variable such as `g:zepl_switchbuf`. If it's set, use its value for `switchbuf`, otherwise use the zepl defaults. That way the user could have a line in their vimrc like `set g:zepl_switchbuf = &switchbuf` to use their own settings.

axvr commented on autoload/zepl.vim:
```diff
@@ -100,6 +100,46 @@ function! zepl#send(text) abort
+    call term_sendkeys(s:repl_bufnr, "\<CR>")
+    call term_wait(s:repl_bufnr)
+  endif
+endfunction
```
These could be simplified.  You mentioned a problem with using `zepl#send` not sending keys I've just pushed some changes which should fix that.  The function now accepts an optional second parameter. If set to `1`, it won't modify the input before sending it to the REPL.  This means that these now works:

```vim
call zepl#send("\<C-d>", 1)
call zepl#send("\<C-c>", 1)
call zepl#send("\<CR>", 1)
```

thezeroalpha replied:
Edit: thanks for changing the function, and please disregard the previous content of this comment.

I found out that the error with mappings (and why I previously 'had to' write three different functions for interrupts and newline) is because Vim would interpret the control characters if they were included in a mapping (for example `nnoremap gzc :<C-u>call zepl#send("\<C-c>", 1)<CR>`), and the mapping would fail. When I pass the control characters to the function by unicode value instead of symbolically (e.g. `\u03` instead of `\<C-c>`), it works.

axvr reviewed (changes requested):
Thank you for the contribution.

I'm actually not sure how (and if) I want to add more features. I quite like the idea of a minimal core, for which users can write extensions for in their vimrc. If I decide not to merge this PR, it'll be because that is the direction I would like to take the plugin.

You highlighted a problem with the `zepl#send` function not sending literal characters.  This has now been fixed (see below). Thank you for bringing this to my attention.

I've requested a few changes below. If I do decide against merging, the resulting code should work from your vimrc (now that I have fixed the `zepl#send` function).

thezeroalpha committed:
Update from master

thezeroalpha committed:
Simplified send functions, not working yet

thezeroalpha committed:
Fixed error with control characters interpreting

I have to pass things like ctrl-c to the function via unicode value, not
symbolically.

thezeroalpha committed:
Merge from origin

thezeroalpha committed:
Adapt to new zepl#jump() and adjust switchbuf

Since there is currently only one REPL, I prefer switching to it even if
it's open in a tab, as opposed to opening a new split.

axvr commented:
Hi @thezeroalpha, I've had a think about this for a while and have eventually come to a decision (sorry it took so long). 

While I think that these changes *could* be useful for some users, I don't think they would be used widely enough to become part of zepl.vim.  I'd also like to keep the number of default key bindings down to a minimum.  So sorry, but I'm going to close this pull request.  As I said in a previous comment this code should now all work as part of your personal Vim configuration.

Thank you for creating this pull request, you highlighted many important issues with the functions, and I welcome any future contributions from you.

thezeroalpha commented:
Hi @axvr, of course, there's nothing to apologise for. It's your project so if you don't think something fits, that's completely fine and your call. Thanks for helping me with the issues I had, and for making this plugin, I've found it very useful in my workflow.

One thing I'd suggest potentially mentioning in the readme/documentation is that if the user wants to send special characters like <cr> or <c-c> to zepl, they may need to use the unicode values -- that took me a while to figure out, and not everyone might think of it.
